### PR TITLE
Convert common types to aliases for easier readability and future refactoring

### PR DESCRIPTION
**Pull request by alonmuroch**
There are a few common types that are very common such as pub keys ([48] bytes), root bytes ([32]byte) and others which is worth refactoring to have aliases.
It reads much easier and any future change will go smoother with much less work.

I've created a const file that includes those types and changed all the go files under validator to use them. Next step will be the beacon folder and then the rest.

### CODE REVIEW

File: shared/params/const.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package params
```
Comment (terencechain) [inline shared/params/const.go]: I'd add comments for each one of them, what is it used?.. etc

**Comment by mcdee** [comment]
Not a fan of these being in `params`, they're more types.

This is also going to play havoc with the 0.10.1/0.11 merge; should this be held off until then?

File: shared/params/const.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+type KeyBytes = [48]byte
```
Comment (mcdee) [inline shared/params/const.go]: This should be `PublicKey` not just `Key`.

File: shared/params/const.go [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+type ValidatorRoleBytes = [48]byte
+type RootBytes = [32]byte
+type AttestationHashBytes = [32]byte
+type BalanceBytes = [48]byte
+type DomainBytes = []byte
+type GraffitiBytes = []byte
+type SoltNumber = uint64
```
Comment (mcdee) [inline shared/params/const.go]: This should be `Slot` not `Solt`

**Comment by terencechain** [comment]
> Not a fan of these being in `params`, they're more types.
> 
> This is also going to play havoc with the 0.10.1/0.11 merge; should this be held off until then?

Same. I dont know I feel about pre defined type in the global level. In some ways, it makes it harder to read/track

**Comment by mcdee** [comment]
Not sure about the naming convention here.  Two things:

  - adding the underlying type (`Slot*Number*`, `Domain*Bytes*`) seems like overkill
  - it would make sense for these to be aligned with the types in the spec at https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/beacon-chain.md#custom-types

There are also some items here that are instances (`AttestationHashBytes`, `BalanceBytes`) rather than types.

**Comment by alonmuroch** [comment]
> Not a fan of these being in `params`, they're more types.
> 
> This is also going to play havoc with the 0.10.1/0.11 merge; should this be held off until then?

Are there any type changes that go along with moving to 0.11?

File: shared/params/const.go (hunk 2 above)
Comment (alonmuroch) [reply shared/params/const.go]: 48 bytes is also the private key no?

File: shared/params/const.go (hunk 3 above)
Comment (alonmuroch) [reply shared/params/const.go]: fixed

File: shared/params/const.go (hunk 1 above)
Comment (alonmuroch) [reply shared/params/const.go]: added

**Comment by alonmuroch** [comment]
> > Not a fan of these being in `params`, they're more types.
> > This is also going to play havoc with the 0.10.1/0.11 merge; should this be held off until then?
> 
> Same. I dont know I feel about pre defined type in the global level. In some ways, it makes it harder to read/track

Why is that? 
There are many types that are very common in the code, why not give them a nice alias in one place?

**Comment by alonmuroch** [comment]
> Not sure about the naming convention here. Two things:
> 
> * adding the underlying type (`Slot*Number*`, `Domain*Bytes*`) seems like overkill
> * it would make sense for these to be aligned with the types in the spec at https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/beacon-chain.md#custom-types
> 
> There are also some items here that are instances (`AttestationHashBytes`, `BalanceBytes`) rather than types.

that makes sense, maybe put them under their own directory called "custom_types" as per the specs + use spec names.
I can make that change NP

**Comment by alonmuroch** [comment]
> I am not sure this is a necessary change, this doesn't increase readability. it does the opposite in my opinion, now I have to always look up whatever the actual values of the types are from the alias

For me it makes things easier but I get where you coming from.
An alternative is to not use aliases but rather constants for byte lengths, so a pub key will be [params.<const length variable of 48>]byte

I prefer aliases

**Comment by rauljordan** [comment]
Hi @alonmuroch in my opinion, this would drastically decrease readability, for a few reasons:

First, Eth2 already has a ton of jargon as is, with the spec full to the brim of nomenclature, "new" notation, and naming that requires you to _constantly_ reference a nomenclature table to successfully understand it. I and a few others also share the sentiment that we constantly have to refer back to the term definitions when trying to read a new function or a PR made into the repo because there is just so much to keep track of. This is akin to using acronyms for everything given they "shorten" long text, but also create further burden for those folks that aren't domain experts.

I would also argue this will reduce contributions from people that are _not_ fully versed in eth2 jargon. The benefit of using barebones types instead of pure aliases for basic things such as uint64 or [32]byte is that any Go developer can come in, see the type signature of a function from something such as Godoc, and understand how to call the function, test it, and comply with its type signature.

Many folks develop with Vim for their day to day, nyself included, for browsing through the repo quickly, and without a real-time IDE that gives type-hints on hover, it becomes harder to understand what a type is if I really need to see it.

I feel like this practice can also lead to a slippery slope of wanting to alias _everything_, even functions, as particular types. It introduces more "magic" into the repo. For example, take the Simple Serialize specification [here](https://github.com/ethereum/eth2.0-specs/blob/dev/ssz/simple-serialize.md). They used to have a ton of `magic` nomenclature such as `List[BytevectorN, BytesN32]` which would not only be inconvenient for most folks, but would further complicate their example codebase which would resort to python magic decorators and aliases for the sake of nomenclature. Since then, they have reduced nomenclature in their specification significantly.  I personally feel this PR is treading the line of introducing Go anti-patterns, as Go strives to be as simple and concise as possible. Being vanilla is a plus here rather than a hindrance.

**Comment by alonmuroch** [comment]
@rauljordan how about we change it to have the byte length for the various types as defined const instead of hardcode numbers? 

Just seems strange to have the same hardcoded types (and their lengths) re-typed everywhere .

**Comment by prestonvanloon** [comment]
@alonmuroch Thanks for your patch. We've discussed this internally and within this PR itself. We feel that this feature does not result in a net benefit for the code readability, quality, or maintainability. Our reasoning here is that aliasing types adds unnecessary cognitive friction compared to primitive go types. While there is no clear right answer or best practice when it comes to type aliasing, the team is unanimously in support of the current use of primitive types.

**Pull request closed without merging**